lib/elixir/lib/enum.ex

```diff
@@ -3214,6 +3214,51 @@ defmodule Enum do
     |> :lists.reverse()
   end
 
+  @doc """
+  Zips two collections together by passing the corresponding element from each collection to the
+  provided fun. Returns a new list that contains the result of that fun for each pair of elements.
+
+  The zipping finishes as soon as any enumerable in the given collection completes.
+
+  ## Examples
+
+      iex> Enum.zip_with([1, 2], [3, 4], fn x, y -> x + y end)
+      [4, 6]
+
+      iex> Enum.zip_with([1, 2], [3, 4, 5, 6], fn x, y -> x + y end)
+      [4, 6]
+
+      iex> Enum.zip_with([1, 2, 5, 6], [3, 4], fn x, y -> x + y end)
+      [4, 6]
+  """
+  @spec zip_with(t, t, (term, term -> term)) :: [term]
+  def zip_with(enumerable1, enumerable2, fun)
+      when is_list(enumerable1) and is_list(enumerable2) do
+    zip_list(enumerable1, enumerable2, fun)
+  end
+
+  def zip_with(enumerable1, enumerable2, fun) do
+    zip_with([enumerable1, enumerable2], fun)
+  end
+
+  @doc """
+  Zips corresponding elements from a finite collection of enumerables into a new enumerable which
+  is built by calling zip_fun with the first element from each enumerable, then the with the
+  second element and so on until any one of the enums runs out of elements.
+
+  ## Examples
+
+      iex> Enum.zip_with([[1, 2], [3, 4], [5, 6]], fn [x, y, z] -> x + y + z end)
+      [9, 12]
+  """
+  @spec zip_with(t, (term, term -> term)) :: [term]
```

Comment:
I think we should defined first zip_with/2 and then zip_with/3

Reply:
Do you mean move the zip_with/2 function above the zip_with/3 definition in this file?

Reply:
Don't worry, the order is fine as is. 👍 

@eksperimental, please don't worry too much about these small details in PRs. I would rather focus on the correctness of the implementation and quality of the documentation, like you did above, than ordering of functions. :)

Reply:
> Do you mean move the zip_with/2 function above the zip_with/3 definition in this file?

Yes, but as José said it does not really matter for now.